python/lsst/meas/algorithms/noise_covariance.py

```diff
@@ -0,0 +1,252 @@
+# This file is part of meas_algorithms.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+import itertools
+from contextlib import contextmanager
+from dataclasses import dataclass
+
+import lsst.afw.image
+import numpy as np
+from lsst.meas.algorithms import SubtractBackgroundTask
+from lsst.pex.config import Config, ConfigurableField, Field, ListField
+from lsst.pipe.base import Task
+
+__all__ = (
+    "ComputeNoiseCorrelationConfig",
+    "ComputeNoiseCorrelationTask",
+    "CorrelationMatrix",
+)
+
+
+@dataclass(frozen=True)
+class CorrelationMatrix:
+    """A class holding correlation coefficients for a set of background pixels.
+
+    CorrelationMatrix is a dataclass that is initialized with a numpy ndarray
+    and provides some convenience methods for accessing the matrix elements.
+    A CorrelationMatrix instance is callable wth two integer values x and y,
+    which returns the <I(m,n) I(m+x, n+y) / sqrt( V(m,n) V(m+x,n+y) )>, where
+    I is the image, V is the variance plane and < > denotes the expectation
+    operator.
+
+    Parameters
+    ----------
+    array : `numpy.ndarray`
+        The matrix of correlation coefficients.
+    """
+
+    array: np.ndarray
+
+    @property
+    def shape(self) -> tuple[int, int]:
+        """The shape of the correlation matrix."""
+        return self.array.shape
+
+    def __call__(self, x: int, y: int) -> float:
+        return self.array[x, y]
+
```

Comment:
It's a simple class for now, but I can see this being extended for e.g, to transform the covariance/correlation matrix at the time of warping etc. in the near future. And the call method is still useful, because with numpy arrays being passed around, there's always the confusion of what the indices mean (vertical first or horizontal first, vertical upwards or downwards etc.). A call method with `x` and `y` arguments makes it clear.